### PR TITLE
feat: implement `ENRResponse` decoding

### LINKED ISSUE #89: Implement `ENRRequest` message decoding

### PR DESCRIPTION
**Pull request by ricomateo**
This PR depends on #137

**Motivation**
Implement discv4 protocol.

**Description**
Enables `ENRResponse` decoding

Closes #89

### CODE REVIEW

File: crates/core/src/rlp/decode.rs [hunk 1]
```diff
@@ -375,6 +375,71 @@ pub fn decode_rlp_item(data: &[u8]) -> Result<(bool, &[u8], &[u8]), RLPDecodeErr
...
 }
 
+/// Splits an RLP item in two:
+/// - The first item including its prefix
+/// - The remaining bytes after the item
+/// It returns a 3-element tuple with the following elements:
+/// - A boolean indicating if the item is a list or not.
+/// - The payload of the item, including its prefix.
+/// - The remaining bytes after the item.
+pub fn get_item_with_prefix(data: &[u8]) -> Result<(bool, &[u8], &[u8]), RLPDecodeError> {
```
Comment (MegaRedHand) [inline crates/core/src/rlp/decode.rs:385]: We aren't using the boolean so we should remove it.

File: crates/core/src/rlp/decode.rs [hunk 2]
```diff
@@ -375,6 +375,71 @@ pub fn decode_rlp_item(data: &[u8]) -> Result<(bool, &[u8], &[u8]), RLPDecodeErr
...
+    let first_byte = data[0];
+
+    match first_byte {
+        0..=0x7F => Ok((false, &data[..1], &data[1..])),
+        0x80..=0xB7 => {
+            let length = (first_byte - 0x80) as usize;
+            if data.len() < length + 1 {
+                return Err(RLPDecodeError::InvalidLength);
+            }
+            Ok((false, &data[0..length + 1], &data[length + 1..]))
```
Comment (MegaRedHand) [inline crates/core/src/rlp/decode.rs:399]: No need to specify the range starts at 0.

File: crates/core/src/rlp/structs.rs [hunk 3]
```diff
@@ -65,6 +64,21 @@ impl<'a> Decoder<'a> {
...
+    pub fn get_encoded_item(self) -> Result<(Vec<u8>, Self), RLPDecodeError> {
+        match get_item_with_prefix(self.payload) {
+            Ok((_is_list, field, rest)) => {
+                let updated_self = Self {
+                    payload: rest,
+                    ..self
+                };
+                Ok((field.to_vec(), updated_self))
+            }
+            Err(_) => Err(RLPDecodeError::Custom(
```
Comment (MegaRedHand) [inline crates/core/src/rlp/structs.rs:77]: Why the custom error? Can't we just propagate this one?

**Review by MegaRedHand — COMMENTED** [review]
Good work!

File: crates/core/src/rlp/decode.rs (hunk 1 above)
Comment (ricomateo) [reply crates/core/src/rlp/decode.rs]: Done!

File: crates/core/src/rlp/decode.rs (hunk 2 above)
Comment (ricomateo) [reply crates/core/src/rlp/decode.rs]: Done!

File: crates/core/src/rlp/structs.rs (hunk 3 above)
Comment (ricomateo) [reply crates/core/src/rlp/structs.rs]: Done!

File: crates/core/src/rlp/decode.rs [hunk 4]
```diff
@@ -375,6 +375,69 @@ pub fn decode_rlp_item(data: &[u8]) -> Result<(bool, &[u8], &[u8]), RLPDecodeErr
     }
 }
 
+/// Splits an RLP item in two:
+/// - The first item including its prefix
+/// - The remaining bytes after the item
+/// It returns a 3-element tuple with the following elements:
+/// - A boolean indicating if the item is a list or not.
+/// - The payload of the item, including its prefix.
+/// - The remaining bytes after the item.
```
Comment (MegaRedHand) [inline crates/core/src/rlp/decode.rs:384]: This doc is outdated

**Review by ElFantasma — APPROVED** [review]
Is it the same as `enr-response-encoding` PR?

**Comment by ricomateo** [comment]
@ElFantasma the `enr-response-encoding` PR included some of these changes because I used this branch as a base.

**Pull request merged**